[Bugfix] Honor --mm_encoder_attn_backend when used

bradleyhd opened the pull request:
Summary:
In https://github.com/vllm-project/vllm/pull/26104, some changes were made in layer.py that resulted in always trying to switch to FA backend for ViT, even when  `VLLM_ATTENTION_BACKEND` is set.

This broke Meta's internal AMD pipelines as it is not desired nor expected behavior. With this change, the models that were changed in the offending PR can explicitly opt-in to this behavior.

Differential Revision: D84946967

DarkLight1337 commented:
Heads up that we have decoupled the two backends in https://github.com/vllm-project/vllm/pull/27061

bradleyhd committed:
honor --mm_encoder_attn_backend when used (#27124)

Summary:
Pull Request resolved: https://github.com/vllm-project/vllm/pull/27124

In https://github.com/vllm-project/vllm/pull/26104, some changes were made in layer.py that resulted in always trying to switch to FA backend for ViT, even when  `VLLM_ATTENTION_BACKEND` is set.

This broke Meta's internal AMD pipelines as it is not desired nor expected behavior. With this change, the models that were changed in the offending PR can explicitly opt-in to this behavior.

Reviewed By: Prowindy

Differential Revision: D84946967

ywang96 committed:
Merge branch 'main' into export-D84946967

ywang96 reviewed (approved):
LGTM - I think we do need to think about how to deal with override in a better way (whether we should honor it truly with the risk of failure or handle the fallback automatically)

ywang96 committed:
Merge branch 'main' into export-D84946967
